github/github.go

```diff
@@ -235,6 +235,37 @@ func NewClient(httpClient *http.Client) *Client {
 	return c
 }
 
+// NewEnterpriseClient returns a new GitHub API client with provided
+// base URL and upload URL (often the same URL).
+// If either URL does not have a trailing slash, one is added automatically.
+// If a nil httpClient is provided, http.DefaultClient will be used.
+//
+// Note that NewEnterpriseClient is a convenience helper only;
+// its behavior is equivalent to using NewClient, followed by setting
+// the BaseURL and UploadURL fields.
+func NewEnterpriseClient(httpClient *http.Client, baseEndpointStr string, uploadEndpointStr string) (*Client, error) {
```

Comment:
When writing idiomatic Go code (see references at the end of this comment), it's common to think about how an identifier will appear in `godoc` and prioritize the readability of the documentation and signature, even if sometimes it requires making small sacrifices to the internal implementation details.

In this case, we have very nice and readable internal variable names `baseURL` and `uploadURL` of type `*url.URL`. Since the function parameter types are different, those names can't be used, so `baseEndpointStr` and `uploadEndpointStr` are used instead.

I suggest we flip that around, and use the nicer `baseURL` and `uploadURL` names in the function signature instead. First of all, it simply reads better. Second, it matches the fields `BaseURL` and `UploadURL` of `Client` struct, so it should be easier to understand they're related. Third, it matches what the documentation says more closely.

Another point, there is a related Go package for accessing the GitHub API v4 [`githubql`](https://github.com/shurcooL/githubql), and it has a similar function already:

https://godoc.org/github.com/shurcooL/githubql#NewEnterpriseClient

I think it'd be better to stay consistent with it and put the `baseURL, uploadURL` parameters first, `httpClient` last. Especially since `httpClient` can be left `nil`, it looks nicer if it's the at the end of the parameter list (but this isn't a big deal). If you have arguments for `httpClient` being better placed first, let me know. We can consider changing `githubql` to match.

So, this is the signature I'd suggest we should go with:

```Go
func NewEnterpriseClient(baseURL, uploadURL string, httpClient *http.Client) (*Client, error)
```

(You can also simplify `baseURL string, uploadURL string` to just `baseURL, uploadURL string`, since they're the same type.)

References:

- https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters
- https://golang.org/doc/effective_go.html#commentary

Reply:
Thanks for the detailed comment, this is a great way for me to learn. I completely agree with your points, also interesting about `godoc` and keeping it in mind when we write our APIs. This is something I will try and do in the future!

I have no strong arguments for keeping `httpClient` first, happy to change this and keep it consistent with the `githubql ` package.

Have added a commit to the branch with the requested changes. Thanks!

Reply:
The parameter names still have previous names rather than `baseURL, uploadURL`. Are you planning on pushing that change in a separate commit?

Reply:
Sorry, you're right, I got ahead of myself there! Added another commit :)